File: labaAlgS2E1/VertexBiconnectedComponent.kt
package labaAlgS2E1

import java.util.*
import kotlin.math.min

object VertexBiconnectedComponent {

    lateinit var edges: Array<MutableList<Int>>
    lateinit var used: BooleanArray
    var t = 0
    lateinit var inTime: IntArray
    lateinit var backEdgeTime: IntArray
    lateinit var cutVertex: BooleanArray
    lateinit var component: IntArray

    @JvmStatic
    fun main(args: Array<String>) {
        val sc = Scanner(System.`in`)
        val n = sc.nextInt()
        val m = sc.nextInt()
        edges = Array(n) { mutableListOf<Int>() }
        used = BooleanArray(n)
        cutVertex = BooleanArray(n)
        t = 0
        inTime = IntArray(n)
        backEdgeTime = IntArray(n)
        component = IntArray(n)

        for (i in 0 until m) {
            val v = sc.nextInt() - 1
            val u = sc.nextInt() - 1
            edges[u].add(v)
            edges[v].add(u)
        }

        for (i in 0 until n) {
            if (!used[i]) {
                if (cutVertexDfs(i, -1) > 1) {
                    cutVertex[i] = true
                }
            }
        }

        for (i in 0 until n) {
            used[i] = false
        }

        var count = 0
        for (i in 0 until n) {
            if (!used[i]) {
                biconnectedComponentdfs(i, -1, ++count)
            }
        }

        println(count)
        for (i in 0 until n) {
            print(component[i])
            print(" ")
        }
    }

    fun cutVertexDfs(v: Int, p: Int): Int {
        var ch = 0
        used[v] = true
        backEdgeTime[v] = t++
        inTime[v] = backEdgeTime[v]
        for (u in edges[v]) {
            if (u != p) {
                if (used[u]) {
                    backEdgeTime[v] = min(backEdgeTime[v], inTime[u])
                } else {
                    cutVertexDfs(u, v)
                    backEdgeTime[v] = min(backEdgeTime[v], backEdgeTime[u])
                    if (backEdgeTime[u] >= inTime[v] && p != -1) {
                        cutVertex[v] = true
                    }
                    ch++
                }
            }
        }
        return ch
    }

    fun biconnectedComponentdfs(v: Int, p: Int, count: Int) {
        used[v] = true
        component[v] = count
        for (u in edges[v]) {
            if (u != p && !used[u]) {
                if (!cutVertex[u]) {
                    biconnectedComponentdfs(u, v, count)
                }
            }
        }
    }

}

File: labaAlgS2E1/VertexBiconnectedComponent.cpp
#include <iostream>
#include <vector>
#include <map>

static std::vector<std::vector<std::pair<int, int>>> edges;
static std::vector<bool> used;
static int t = 0;
static std::vector<int> inTime;
static std::vector<int> backEdgeTime;
static std::vector<bool> ans;

static int maxComponent = 0;
static std::vector<int> component;

void paint(int v, int color, int e) {
    used[v] = true;
    for (std::pair<int,int> pair : edges[v]) {
        int u = pair.first;
        int k = pair.second;
        if (k != e) {
            if (!used[u]) {
                if (backEdgeTime[u] >= inTime[v]) {
                    component[k] = ++maxComponent;
                    paint(u, maxComponent, k);
                } else {
                    component[k] = color;
                    paint(u, color, k);
                }
            } else if (inTime[u] < inTime[v]) {
                component[k] = color;
            }
        }
    }
}

void dfs(int v, int e) {
    used[v] = true;
    inTime[v] = backEdgeTime[v] = t++;;
    for (std::pair<int,int> pair : edges[v]) {
        int u = pair.first;
        int k = pair.second;
        if (k != e) {
            if (used[u]) {
                backEdgeTime[v] = std::min(backEdgeTime[v], inTime[u]);
            } else {
                dfs(u, k);
                backEdgeTime[v] = std::min(backEdgeTime[v], backEdgeTime[u]);
            }
        }
    }
}

int main() {
    int n, m;
    std::cin >> n >> m;
    edges.resize(n);
    used.assign(n, 0);
    ans.assign(n, 0);
    t = 0;
    inTime.assign(n, 0);
    backEdgeTime.assign(n, 0);
    component.assign(m, 0);

    for (int i = 0; i < m; ++i) {
        int v, u;
        std::cin >> v >> u;
        edges[u - 1].push_back(std::make_pair(v - 1, i));
        edges[v - 1].push_back(std::make_pair(u - 1, i));
    }

    for (int i = 0; i < n; ++i) {
        if (!used[i]) {
            dfs(i, -1);
        }
    }
    used.assign(n, false);
    for (int i = 0; i < n; ++i) {
        if (!used[i]) {
            paint(i, maxComponent, -1);
        }
    }

    std::cout << maxComponent << std::endl;
    for (int i = 0; i < m; ++i) {
        std::cout << component[i] << " ";
    }
    return 0;
}
